### ISSUE TITLE
Direction for this plugin

### ISSUE THREAD

**Issue by jmezach** [benwilcock/backstage-plugin-backchat#1]
First of all, thanks a lot for this PoC @benwilcock. I've managed to get it running on my own machine and it works pretty well. As I mentioned offline at KubeCon I really like this idea of having an AI powered Chatbot within Backstage. I've PoC'ed something similar within my company based on Azure OpenAI and I think this is something that we're really interested in, especially since you're running the AI within your own private space which ensures that you don't accidentally leak company secrets into the public ones. So I'm keen on pushing this idea further and would love to help out.

Right now this plugin is essentially re-hosting the UI of either ChatbotUI and/or Text Gen Web UI using an iframe. While that works well for a proof of concept I don't think it is very feasible in an actual production deployment, let alone the security implications of such a deployment where you're exposing those tools directly to the end user.

So I think for this to be truly viable I think we'd have to build the UI into the plugin itself. If we're going that route I think it makes sense to all add a backend plugin that could save the conversations and provide a pluggable way of integrating with various backends to provide the actual AI if that makes sense.

Any thoughts on this?

**Comment by benwilcock** [comment]
Absolutely. I would love it to be a proper client with its own gui. That's certainly the direction I was hoping we could take. I also really would like to keep the idea of having either changeable backends or changeable models (or both) so that folks can choose the model they need and run with it. 

I'm so glad you like the idea. It would be great if you could bring your skills to beat on it to turn the dream into a reality.

**Comment by jmezach** [comment]
I've done some early prototyping of the UI and I would love to get some feedback on this.

![Web capture_13-11-2023_145719_localhost](https://github.com/benwilcock/backstage-plugin-backchat/assets/1225489/dc43d9d2-a8f1-498b-ba77-1fd84e92570b)

**Comment by benwilcock** [comment]
Seems like a good start. As far as the UI/UX goes, I think the overarching look and feel of [ChatbotUI](https://www.chatbotui.com/) is probably a good analogue as it allows users to do 4 main tasks well:

* Chat
* Store and recall chats
* Store and recall prompts
* Change and configure models (e.g temperature).

This is a great start towards that. Keep going!

**Comment by jmezach** [comment]
Yes, this was definitely inspired by ChatbotUI, just build up with Backstage/Material-UI components. There's definitely more finetuning to be done here, but I guess it's a start. I've committed the things I have so far to the `main` branch in my fork. I've also introduced a feature flag so that it can be easily switched back to the current version.

So far I've build everything into the BackchatComponent directly, so that will need some refactoring. I'll also have to look into how I'm going to add a backend plugin to the repo as well.

**Comment by benwilcock** [comment]
I think when I looked into backend plugins in the past, they're either separate repos, or some sort of folder structure is introduced where they're siblings like this example from the QetA plugin: [https://github.com/drodil/backstage-plugin-qeta/tree/main/plugins](https://github.com/drodil/backstage-plugin-qeta/tree/main/plugins).

Would it be more convenient for me to add you as a contributor so you don't need to maintain a fork?

**Comment by jmezach** [comment]
Yeah, I've already done that restructuring on my fork which seems to work on my machine at least so far. Somehow the backend plugin I've added isn't working though and I'm not entirely sure why yet. I've dropped a question in the Backstage Discord to figure out what I've done wrong ;).

And yes, maybe it would be more convenient to be a contributor here.

**Comment by benwilcock** [comment]
Today I added another switchable GUI and updated all the documentation in both repos. I got excited by Ollama Web UI's simplicity and ease of use. It's great to look at and it offers a very good downloader for LLMs. There's just one problem - I don't think it's OpenAI API compatible - which is a shame. If it had OpenAI API it would be my new favorite for sure.

We should probably talk roadmap. I'm thinking that there's a possibility we don't need a backend plugin.
